bitsail-connectors/connector-druid/src/main/java/com/bytedance/bitsail/connector/druid/sink/DruidWriter.java

```diff
@@ -0,0 +1,267 @@
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void write(final Row element) {
+    final StringJoiner joiner = new StringJoiner(DEFAULT_FIELD_DELIMITER, "", "");
+    for (int i = 0; i < element.getArity(); i++) {
+      final Object v = element.getField(i);
+      if (v != null) {
+        joiner.add(v.toString());
+      }
+    }
+    // timestamp column is a required field to add in Druid.
+    // See https://druid.apache.org/docs/24.0.0/ingestion/data-model.html#primary-timestamp
+    joiner.add(String.valueOf(processTime));
+    this.data.append(joiner);
+    this.data.append(DEFAULT_LINE_DELIMITER);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void flush(final boolean endOfInput) throws IOException {
+    final ParallelIndexIOConfig ioConfig = provideDruidIOConfig(this.data);
+    final ParallelIndexSupervisorTask indexTask = provideIndexTask(ioConfig);
+    final String inputJSON = provideInputJSONString(indexTask);
+    final byte[] input = inputJSON.getBytes();
+    try (final OutputStream os = httpURLConnection.getOutputStream()) {
+      os.write(input, 0, input.length);
+    }
+    try (final BufferedReader br =
+                 new BufferedReader(new InputStreamReader(httpURLConnection.getInputStream(), StandardCharsets.UTF_8))) {
+      final StringBuilder response = new StringBuilder();
+      String responseLine;
+      while ((responseLine = br.readLine()) != null) {
+        response.append(responseLine.trim());
+      }
+      log.info("Druid write task has been sent, and the response is {}", response);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public List<CommitT> prepareCommit() {
+    return null;
+  }
+
+  private HttpURLConnection provideHttpURLConnection(final String coordinatorURL) throws IOException {
+    final URL url = new URL("http://" + coordinatorURL + DRUID_ENDPOINT);
+    final HttpURLConnection con = (HttpURLConnection) url.openConnection();
+    con.setRequestMethod("POST");
+    con.setRequestProperty("Content-Type", "application/json");
+    con.setRequestProperty("Accept", "application/json, text/plain, */*");
+    con.setDoOutput(true);
+    return con;
+  }
+
+  private ObjectMapper provideDruidSerializer() {
+    final ObjectMapper mapper = new ObjectMapper();
+    mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+    mapper.configure(MapperFeature.AUTO_DETECT_GETTERS, false);
+    mapper.configure(MapperFeature.AUTO_DETECT_FIELDS, false);
+    mapper.configure(MapperFeature.AUTO_DETECT_IS_GETTERS, false);
+    mapper.configure(MapperFeature.AUTO_DETECT_SETTERS, false);
+    mapper.configure(SerializationFeature.INDENT_OUTPUT, false);
+    mapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false);
+    return mapper;
+  }
+
+  private DataSchema provideDruidDataSchema(final BitSailConfiguration writerConfig) {
+    final String datasource = writerConfig.getNecessaryOption(DATASOURCE, REQUIRED_VALUE);
```

Comment:
based on the error message, I wouldn't know which necessary option I am missing. Can we find a way to expose this information?

Reply:
The [**getNecessaryOption** method](https://github.com/bytedance/bitsail/blob/master/bitsail-common/src/main/java/com/bytedance/bitsail/common/configuration/BitSailConfiguration.java#L192) prints out DATASOURCE in error message. I think it's good enough. Or do you have different thought?

Reply:
I see, that's sufficient enough